src/twisted/internet/endpoints.py

```diff
@@ -847,7 +866,7 @@ def __init__(
         """
 
         self._reactor = reactor
-        self._nameResolver = self._getNameResolverAndMaybeWarn(reactor)
+        self._getNameResolverAndMaybeWarn(reactor)
```

Comment:
Not sure why need this here.

I tried to add a comment based on my understanding.

I don't really know why we need a different resolver for each connection.

I think that the resolver change could be moved to a separate ticket/PR.

```suggestion
        # This is called here to validate the reactor.
        # We will get the resolver later at connection time
        # so that if we reuse the endpoint for multiple connection
        # we get a new resolve for each connection.
        self._getNameResolverAndMaybeWarn(reactor)
```

Reply:
Huh. Thanks. This is a really good catch. I seem to have introduced a really obscure TOCTOU bug here for… no reason. Probably because i was fighting some tests and trying to inject the name-resolution dependency at a later point.

Reply:
OK, trying to resolve this, I got some test failures, and realized that this is actually a bugfix; we are not getting a "new resolver for each connection", but rather, allowing the pluggable name resolver in the reactor to actually be modified. As it currently stands, the hostname endpoint is eagerly fetching the name resolver from the reactor when it is constructed, then never honoring any modifications that the application might make to the reactor, which is just kind of wrong.  I will add a comment that describes this.

(In principle, the resolver changes should have been a separate PR, but at this point, separating them is probably more trouble than it's worth.)